File: bench/bench_ocaml/lib/math.ml
open Common

(** This function is the integer square root function. *)
let isqrt (y : nat) =
  if y > three then
    let z = y in
    let x = (y / two) + Z.one in
    let rec iter ((x, y, z) : nat * nat * nat) : nat =
      if x < z then iter (((y / x) + x) / two, y, x) else z
    in
    iter (x, y, z)
  else if y <> Z.zero then Z.one
  else Z.zero

let rec fast_power : nat -> nat -> nat -> nat =
 fun a b result ->
  if b = Z.zero then result
  else
    let result = if b land Z.one = Z.one then result * a else result in
    let b = b lsr 1 in
    let a = a * a in
    fast_power a b result

(** return x ^ y *)
let power ((x, y) : nat * nat) : nat = fast_power x y Z.one

(** It computes the factorial n! for a given nat n. (i.e. n*(n-1)*(n-2)*...*1n ) *)
let factorial (n : nat) : nat =
  let rec fact ((acc, i) : nat * nat) : nat =
    if i < two then acc else fact (acc * i, abs (i - Z.one))
  in
  fact (Z.one, n)

let min (a : nat) (b : nat) : nat = if a < b then a else b
let max (a : nat) (b : nat) : nat = if a > b then a else b

(* // TODO : optimize log algorithm ! *)
let log_10 (x : nat) : nat =
  let rec check_power ((x, i) : nat * nat) : nat =
    if x mod power (ten, i) > Z.zero then abs (i - Z.one)
    else check_power (x, i + Z.one)
  in
  check_power (x, Z.one)

File: bench/bench_kotlin/src/main/kotlin/math/Math.kt
package math.lib

import java.math.BigInteger
import java.math.BigInteger.ONE
import java.math.BigInteger.TWO
import java.math.BigInteger.ZERO

object Math {

    val EIGHTEEN: BigInteger = BigInteger.valueOf(18)

    fun iSqrt(y: BigInteger): BigInteger =
        if (y > BigInteger.valueOf(3)) {
            tailrec fun iter(x: BigInteger, y: BigInteger, z: BigInteger): BigInteger =
                if (x < z)
                    iter(((y / x) + x) / TWO, y, x)
                else
                    z

            iter((y / TWO) + ONE, y, y)
        } else if (y != ZERO) {
            ONE
        } else {
            ZERO
        }

    @Suppress("NAME_SHADOWING")
    fun power(x: BigInteger, y: BigInteger): BigInteger {
        tailrec fun multiply(a: BigInteger, b: BigInteger, result: BigInteger): BigInteger =
            if (b == ZERO) {
                result
            } else {
                val result = if (b.and(ONE) == ONE) result * a else result
                val b = b.shr(1);
                val a = a * a
                multiply(a, b, result)
            }

        return multiply(x, y, ONE)
    }

    fun factorial(n: BigInteger): BigInteger {
        tailrec fun factorial(acc: BigInteger, i: BigInteger): BigInteger =
            if (i < TWO) {
                acc
            } else {
                factorial(acc * i, i - ONE)
            }

        return factorial(ONE, n)
    }

    fun min(a: BigInteger, b: BigInteger): BigInteger =
        if (a < b) {
            a
        } else {
            b
        }

    fun max(a: BigInteger, b: BigInteger): BigInteger =
        if (a < b) {
            b
        } else {
            a
        }

    fun log10(x: BigInteger): BigInteger {
        tailrec fun checkPower(x: BigInteger, i: BigInteger): BigInteger =
            if (x % power(BigInteger.TEN, i) > ZERO) {
                (i - ONE).abs()
            } else {
                checkPower(x, i + ONE)
            }

        return checkPower(x, ONE)
    }
}